File: Lv 3/N으로 표현.py
from collections import deque


def solution(N, number):
    answer = -1
    table = []
    for i in range(1, 9):
        middle_set = set()
        middle_set.add(int(str(N) * i)) # 쭉 붙인거는 따로 추가
        for j in range(0, i-1):
            for k in table[j]:
                for l in table[-j-1]:
                    middle_set.add(k + l)
                    middle_set.add(k - l)
                    middle_set.add(k * l)
                    if l != 0:
                        middle_set.add(k // l)
        if number in middle_set:
            return i
        table.append(middle_set)

            
    return answer

File: Lv 3/N으로 표현.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class Solution {
    public int solution(int N, int number) {
        int answer = 0;
        List<Set<Long>> table = new ArrayList<>();
        table.add(new HashSet<>());
        for (int i = 1; i <= 8; i++) {
            Set<Long> numbers = new HashSet<>();
            numbers.add(Long.parseLong((""+N).repeat(i)));
            for (int j = 1; j < i; j++) {
                Set<Long> source = table.get(j);
                Set<Long> target = table.get(i-j);
                for (long sourceValue : source) {
                    for (long targetValue : target) {
                        numbers.add(sourceValue + targetValue);
                        numbers.add(sourceValue - targetValue);
                        numbers.add(sourceValue * targetValue);
                        if (targetValue != 0) {
                            numbers.add(sourceValue / targetValue);
                        }
                    }
                }
            }
            for (long num : numbers) {
               if (num == number) {
                   return i;
               }
            }
            table.add(numbers);
        }
        return -1;
    }
}
